src/Workspaces/Remote/ServiceHub/Shared/RemoteEndPoint.cs

```diff
@@ -197,82 +210,114 @@ private static bool ReportUnlessCanceled(Exception ex, CancellationToken remoteT
             }
 
-            if (remoteToken.IsCancellationRequested)
+            if (linkedCancellationToken.IsCancellationRequested)
             {
-                // now we allow connection to be closed by users by killing remote host process.
-                // in those case, it will be converted to remote token cancellation. we accept that as known
-                // exception, and allow us to not crash
+                // Connection can be closed when the remote process is killed.
+                // That will manifest as remote token cancellation.
                 return true;
             }
 
-            // unexpected exception case. crash VS
-            return FatalError.Report(ex);
-        }
-
-        private static void RaiseCancellationIfInvokeFailed(Task task, CancellationTokenSource mergedCancellation, CancellationToken cancellationToken)
-        {
-            // if invoke throws an exception, make sure we raise cancellation
-            var dummy = task.ContinueWith(p =>
-            {
-                try
-                {
-                    // now, we allow user to kill OOP process, when that happen, 
-                    // just raise cancellation. 
-                    // otherwise, stream.WaitForDirectConnectionAsync can stuck there forever since
-                    // cancellation from user won't be raised
-                    mergedCancellation.Cancel();
-                }
-                catch (ObjectDisposedException)
-                {
-                    // merged cancellation is already disposed
-                }
-            }, cancellationToken, TaskContinuationOptions.NotOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+            ReportNonFatalWatson(ex, UnexpectedExceptionLogMessage);
+            return true;
         }
 
         private bool ReportUnlessCanceled(Exception ex, CancellationToken cancellationToken)
         {
-            if (cancellationToken.IsCancellationRequested)
+            if (!cancellationToken.IsCancellationRequested)
             {
-                // let cancellation exception thru if they are associated with our cancellation token
-                return false;
+                ReportNonFatalWatson(ex, UnexpectedExceptionLogMessage);
             }
 
-            _logger.TraceEvent(TraceEventType.Error, 1, ex.ToString());
+            return true;
+        }
 
+        private void ReportNonFatalWatson(Exception ex, string message)
+        {
             s_debuggingLastDisconnectReason = _debuggingLastDisconnectReason;
             s_debuggingLastDisconnectCallstack = _debuggingLastDisconnectCallstack;
 
-            // send NFW to figure out why this is happening
-            ex.ReportServiceHubNFW("RemoteHost Failed");
+            ReportNonFatalWatsonWithServiceHubLogs(ex, message);
 
             GC.KeepAlive(_debuggingLastDisconnectReason);
             GC.KeepAlive(_debuggingLastDisconnectCallstack);
```

Comment:
do we need these if they're stored in statics?

Reply:
I don't think so.